app/models/snapshot.rb

```diff
@@ -0,0 +1,286 @@
+  end
+
+  def query_advisers_part_of_institute_financial_planning
+    advisers_part_of('Institute of Financial Planning')
+  end
+
+  def query_advisers_part_of_institute_financial_services
+    advisers_part_of('Institute of Financial Services')
+  end
+
+  def query_advisers_part_of_ci_bankers_scotland
+    advisers_part_of('The Chartered Institute of Bankers in Scotland')
+  end
+
+  def query_advisers_part_of_ci_securities_and_investments
+    advisers_part_of('The Chartered Institute for Securities and Investments')
+  end
+
+  def query_advisers_part_of_cfa_institute
+    advisers_part_of('CFA Institute')
+  end
+
+  def query_advisers_part_of_chartered_accountants
+    advisers_part_of('Institute of Chartered Accountants for England and Wales')
+  end
+
+  def run_queries_and_save
+    run_queries
+    save
+  end
+
+  private
+
+  def advisers_in_country(advisers, country)
+    postcodes = advisers.map { |adviser| adviser.postcode }
+    country_postcodes = country_postcodes(postcodes, country)
+    advisers.select { |adviser| country_postcodes.include?(adviser.postcode) }
+  end
+
+  def advisers_part_of(professional_body)
+    Adviser.includes(:professional_bodies).where(professional_bodies: { name: professional_body })
+  end
+
+  def advisers_who_travel(distance)
+    Adviser.where(travel_distance: TravelDistance.all[distance])
+  end
+
+  def advisers_with_accreditation(accreditation)
+    Adviser.includes(:accreditations).where(accreditations: { name: accreditation })
+  end
+
+  def advisers_with_qualification(qualification)
+    Adviser.includes(:qualifications).where(qualifications: { name: qualification })
+  end
+
+  def country_postcodes(postcodes, country)
+    map_postcodes_to_country(postcodes)
+      .select { |postcode, postcode_country| postcode_country == country }
+      .map { |postcode, postcode_country| postcode }
+  end
+
+  def firms_in_country(firms, country)
+    postcodes = firms.map { |firm| firm.main_office.address_postcode }
+    country_postcodes = country_postcodes(postcodes, country)
+    firms.select { |firm| country_postcodes.include?(firm.main_office.address_postcode) }
+  end
+
+  # Make sure we only request 100 at a time
+  def map_postcodes_to_country(postcodes)
+    postcodes
+      .uniq
+      .each_slice(100)
+      .map { |slice| map_postcodes_slice_to_country(slice) }
+      .reduce(&:merge)
+  end
+
+  def map_postcodes_slice_to_country(postcodes)
+    request = Net::HTTP::Post.new('/postcodes')
+    request.set_form_data(postcodes: postcodes)
+
+    response = Net::HTTP.new('api.postcodes.io').request(request)
+
+    if response.code.to_i == 200
+      result = JSON.parse(response.read_body)['result'].map { |r| r['result'] }.compact
+      result.each_with_object({}) do |r, obj|
+        obj[r['postcode']] = r['country']
+      end
+    else
+      {}
+    end
+  end
+
+  def publishable_firms
+    @_publishable_firms ||= Firm.registered.select(&:publishable?)
+  end
+
+  def run_queries
```

Comment:
Maybe put a comment beside this to explain the intent (I understand because I've just chatted to you about it)